### PR TITLE
Multithread Queue and Basic Project Structure

### PR DESCRIPTION
**Pull request by lietzdaniel**
Implemented Queue with a Mutex on Write Operations, which will be used for the multithreaded Stock Simulation later on.

Also added basic project structure:
- tests folder with tests using Google Test, 
- a utils folder for the threaded Queue called TQueue and possible other Utils coming in the future,
- src folder for the Main File

Google Test is incorporated in the CMakeLists.txt and is automatically fetched. Implemented some Tests for the TQueue.
Brief instruction on how to run the tests are in the README.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -30,3 +30,6 @@
 *.exe
 *.out
 *.app
+
+.vscode/*
+build/*
```
Comment (noakanois) [inline .gitignore]: new line at the end of file

File: README.md [hunk 2]
```diff
@@ -1,2 +1,12 @@
 # stock-simulator
 Application simulating a stock data feed.
+
+
+## Running the Program
+
+1. `mkdir build`
```
Comment (noakanois) [inline README.md]: this shouldnt be in the steps to do it, should autogenerate from the make or cmake or include it

File: src/main.cpp [hunk 3]
```diff
@@ -0,0 +1,6 @@
+#include "main.h"
+
+int main(int argc, char *argv[])
+{
+   return 1;
+}
```
Comment (noakanois) [inline src/main.cpp]: New line at the end of file

File: src/main.h [hunk 4]
```diff
@@ -0,0 +1 @@
+int main(int argc, char *argv[]);
```
Comment (noakanois) [inline src/main.h]: new line

File: tests/tqueue_general_tests.cpp [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    EXPECT_EQ(t_queue.back(), "!");
+    t_queue.pop();
+    EXPECT_EQ(t_queue.back(), "!");
+    t_queue.pop();
+}
+
+int main(int argc, char** argv) {
+    ::testing::InitGoogleTest(&argc, argv);
+    return RUN_ALL_TESTS();
+}
```
Comment (noakanois) [inline tests/tqueue_general_tests.cpp]: new line

File: utils/tqueue/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,5 @@
+cmake_minimum_required(VERSION 3.20.0)
+project(TQueue)
+
+add_library(TQueue INTERFACE)
+target_include_directories(TQueue INTERFACE ${CMAKE_CURRENT_SOURCE_DIR})
```
Comment (noakanois) [inline utils/tqueue/CMakeLists.txt]: new line

File: utils/tqueue/tqueue.h [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+    tqueue<T>& operator=( const tqueue<T>& other){
+        this->queue_write_mutex.lock();
+        this->internal_queue = other;
+        this->queue_write_mutex.unlock();
+    }
+    
+};
+
+
+
```
Comment (noakanois) [inline utils/tqueue/tqueue.h]: use a formatter

File: src/main.h (hunk 4 above)
Comment (noakanois) [inline src/main.h]: Should we have import guards in here?

File: tests/tqueue_general_tests.cpp [hunk 8]
```diff
@@ -0,0 +1,101 @@
+#include <gtest/gtest.h>
+#include "tqueue.h"
```
Comment (noakanois) [inline tests/tqueue_general_tests.cpp]: I would say choose one way and stick with it

File: tests/tqueue_general_tests.cpp [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+
+TEST(TQueueTest, Constructor) {
+    EXPECT_NO_THROW(tqueue<std::string> t_queue;);
+    EXPECT_NO_THROW(tqueue<int> t_queue;);
+    EXPECT_NO_THROW(tqueue<char> t_queue;);
+}
+
+TEST(TQueueTest, Empty) {
```
Comment (noakanois) [inline tests/tqueue_general_tests.cpp]: The testname is non descriptive

File: tests/tqueue_general_tests.cpp [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+    t_queue.pop();
+    EXPECT_EQ(t_queue.size(), 2);
+    t_queue.pop();
+    EXPECT_EQ(t_queue.size(), 1);
+    t_queue.pop();
+    EXPECT_EQ(t_queue.size(), 0);
+}
+
+
+TEST(TQueueTest, PushPop) {
```
Comment (noakanois) [inline tests/tqueue_general_tests.cpp]: The testname is non descriptive


you can test PushPop in 20 different ways

File: tests/tqueue_general_tests.cpp [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+    t_queue.pop();
+    EXPECT_EQ(t_queue.front(), "!");
+    EXPECT_EQ(t_queue.back(), "!");
+    EXPECT_EQ(t_queue.size(), 1);
+    t_queue.pop();
+    EXPECT_EQ(t_queue.size(), 0);
+}
+
+
+TEST(TQueueTest, front) {
```
Comment (noakanois) [inline tests/tqueue_general_tests.cpp]: The testname is non descriptive

File: tests/tqueue_general_tests.cpp [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+    t_queue.push("!");
+    EXPECT_EQ(t_queue.front(), "Hello");
+    t_queue.pop();
+    EXPECT_EQ(t_queue.front(), "World");
+    t_queue.pop();
+    EXPECT_EQ(t_queue.front(), "!");
+    t_queue.pop();
+}
+
+TEST(TQueueTest, back) {
```
Comment (noakanois) [inline tests/tqueue_general_tests.cpp]: The testname is non descriptive

File: utils/tqueue/tqueue.h [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+    void pop(){
+        this->queue_write_mutex.lock();
+        this->internal_queue.pop();
+        this->queue_write_mutex.unlock();
+    }
+
+
+    void swap( tqueue& other ){
+        this->queue_write_mutex.lock();
+        this->internal_queue.swap();
```
Comment (noakanois) [inline utils/tqueue/tqueue.h:75]: you need to put the other que in the .swap() right?

File: README.md (hunk 2 above)
Comment (lietzdaniel) [reply README.md]: This was meant as a very preliminary README just to have this here. However I agree that this is essential and I will add the Makefile for this in this PR already.

File: utils/tqueue/tqueue.h [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+
+
+    void push_range( T&& rg ){
+        this->queue_write_mutex.lock();
+        this->internal_queue.push_range(rg);
+        this->queue_write_mutex.unlock();
+    }
+
+
+    decltype(auto) emplace( T&& t ){
```
Comment (noakanois) [inline utils/tqueue/tqueue.h]: what is this decltype about?

File: utils/tqueue/tqueue.h (hunk 7 above)
Comment (lietzdaniel) [reply utils/tqueue/tqueue.h]: I got a formatter and added a New Line to all Files where this was a problem.

File: src/main.h (hunk 4 above)
Comment (lietzdaniel) [reply src/main.h]: Yes, we should. Implemented.

File: utils/tqueue/tqueue.h [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+    std::queue<T> internal_queue;
+    std::mutex queue_write_mutex;
+
+    public: 
+
+    tqueue() : internal_queue(), queue_write_mutex() {}
+
+
+    T& front(){
+        return this->internal_queue.front();
```
Comment (noakanois) [inline utils/tqueue/tqueue.h:15]: Reader functions need mutexes too i would say, what if you call front while another thread deletes the item

File: utils/tqueue/tqueue.h (hunk 14 above)
Comment (lietzdaniel) [reply utils/tqueue/tqueue.h]: This is taken from the signature of the function of the standard library to wrap it around for TQueue. https://en.cppreference.com/w/cpp/container/queue/emplace
Do you suggest changing it or just using void here? I agree that decltype(auto) is not very pretty.

Comment (noakanois) [reply utils/tqueue/tqueue.h]: ah right, didnt know emplace returns that since C++17

Comment (noakanois) [reply utils/tqueue/tqueue.h]: keep it then

File: README.md (hunk 2 above)
Comment (lietzdaniel) [reply README.md]: I created a Makefile that executes the order of events when running `make`.

File: tests/tqueue_general_tests.cpp (hunk 8 above)
Comment (lietzdaniel) [reply tests/tqueue_general_tests.cpp]: Done, decided for "

File: utils/tqueue/tqueue.h (hunk 15 above)
Comment (lietzdaniel) [reply utils/tqueue/tqueue.h:15]: Yes. What do you think about empty and size, should the mutex be set there too? I think we should just do all the functions.

File: utils/tqueue/tqueue.h (hunk 13 above)
Comment (lietzdaniel) [reply utils/tqueue/tqueue.h:75]: Yes

File: utils/tqueue/tqueue.h (hunk 15 above)
Comment (noakanois) [reply utils/tqueue/tqueue.h:15]: Yes. They also need it.

**Pull request closed without merging**